File: queries.go
package main

import (
	"context"
	"encoding/json"
	"fmt"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type Customer struct {
	Name         string `bson:"name"`
	EmailAddress string `bson:"email_address"`
}

func main() {
	client, _ := mongo.Connect(context.TODO(), options.Client().ApplyURI("mongodb://localhost:27017"))
	defer client.Disconnect(context.TODO())
	database := client.Database("mydb")
	database.CreateCollection(context.TODO(), "customers")
	collection := database.Collection("customers")

	document := Customer{Name: "Charlie", EmailAddress: "charlie@gmail.com"}
	collection.InsertOne(context.TODO(), document)

	document = Customer{Name: "Bob", EmailAddress: "bob@gmail.com"}
	collection.InsertOne(context.TODO(), document)

	document = Customer{Name: "Alice", EmailAddress: "alice@outlook.com"}
	collection.InsertOne(context.TODO(), document)

	searchQuery := bson.D{{Key: "email_address", Value: "bob@gmail.com"}}
	var customer bson.M
	// collection.FindOne(context.TODO(), searchQuery).Decode(&customer)
	// retrieves the first result only
	// collection.Distinct(context.TODO(), "name", bson.D{}) retrieves all distinct names
	// searchQuery := bson.D{{Key: "email_address", Value: bson.D{{Key: "$regex", Value: "^bob@"}}}}
	// retrieves all documents with email addresses starting with "bob@"
	cursor, _ := collection.Find(context.TODO(), searchQuery)
	cursor.Next(context.TODO())
	cursor.Decode(&customer)
	fmt.Println(customer)
	jsonbytes, _ := json.Marshal(customer)
	fmt.Println(string(jsonbytes))
	// use _, hasField := customer["field"] to see if a field exists (hasField is a bool)
	fmt.Println(customer["name"])
	fmt.Println(customer["email_address"])

	searchQuery = bson.D{{Key: "email_address", Value: "alice@outlook.com"}}
	updatedField := bson.D{{Key: "email_address", Value: "alice@gmail.com"}}
	updater := bson.D{{Key: "$set", Value: updatedField}}
	collection.UpdateMany(context.TODO(), searchQuery, updater)

	searchQuery = bson.D{{Key: "email_address", Value: "charlie@gmail.com"}}
	collection.DeleteMany(context.TODO(), searchQuery)

	// use Value: -1 to sort in reverse (descending) order
	opts := options.Find().SetSort(bson.D{{Key: "name", Value: 1}})
	cursor, _ = collection.Find(context.TODO(), bson.D{}, opts)
	var res []bson.M
	cursor.All(context.TODO(), &res)
	for _, document := range res {
		fmt.Println(document)
	}

	count, _ := collection.CountDocuments(context.TODO(), bson.D{})
	fmt.Println(count)
}

File: queries.js
const mongodb = require("mongodb");

const client = new mongodb.MongoClient("mongodb://localhost:27017");
const database = client.db("mydb");

async function run() {
    try {
        await database.createCollection("customers");
        const collection = database.collection("customers");

        let document = {name: "Charlie", email_address: "charlie@gmail.com"};
        await collection.insertOne(document);

        document = {name: "Bob", email_address: "bob@gmail.com"};
        await collection.insertOne(document);

        document = {name: "Alice", email_address: "alice@outlook.com"};
        await collection.insertOne(document);

        let searchQuery = {email_address: "bob@gmail.com"};
        // await collection.findOne(searchQuery) retrieves the first result only
        // await collection.distinct("name") retrieves all distinct names
        // collection.find({email_address: /^bob@/}) retrieves all documents
        // with email addresses starting with "bob@"
        let cursor = collection.find(searchQuery);
        const customer = await cursor.next();
        console.log(customer);
        console.log(JSON.stringify(customer));
        // use customer.hasOwnProperty("field") to see if a field exists
        console.log(customer.name);
        console.log(customer.email_address);

        searchQuery = {email_address: "alice@outlook.com"};
        const updatedField = {email_address: "alice@gmail.com"};
        const updater = {$set: updatedField};
        await collection.updateMany(searchQuery, updater);

        searchQuery = {email_address: "charlie@gmail.com"};
        await collection.deleteMany(searchQuery);

        // use "descending" to sort in reverse order
        cursor = collection.find().sort("name", "ascending");
        while (await cursor.hasNext()) {
            console.log(await cursor.next());
        }

        console.log(await collection.countDocuments());
    } finally {
        await client.close();
    }
}
run();
